### DISCUSSION TITLE
We have gui apps, but when will we be able to run gnome-session?

### DISCUSSION THREAD

**Discussion by Guarmanda** [microsoft/wslg#81] [Q&A]
Some days ago, I was like "I'll totally remove windows and switch to windows". But wslg opens a new road. I'd like to know when we will be able to run an entire gnome session and not just a gui application?

It would be for me and many users a way to totally forget about dual-boot, I mean, we would have linux AND  windows booted at the same time, no more restarting between os, copying documents from linux to windows without having to reboot on windows, having a complete graphic environment... 

```
gnome-session-binary[657]: WARNING: Using null backend for session tracking
GNOME_KEYRING_CONTROL=/root/.cache/keyring-MY4A20
GNOME_KEYRING_CONTROL=/root/.cache/keyring-MY4A20
SSH_AUTH_SOCK=/root/.cache/keyring-MY4A20/ssh
GNOME_KEYRING_CONTROL=/root/.cache/keyring-MY4A20
SSH_AUTH_SOCK=/root/.cache/keyring-MY4A20/ssh
Window manager warning: Unsupported session type
gnome-session-binary[657]: WARNING: App 'org.gnome.Shell.desktop' exited with code 1
Window manager warning: Unsupported session type
gnome-session-binary[657]: WARNING: App 'org.gnome.Shell.desktop' exited with code 1
gnome-session-binary[657]: WARNING: App 'org.gnome.Shell.desktop' respawning too quickly
gnome-session-binary[657]: CRITICAL: We failed, but the fail whale is dead. Sorry....
```

**Comment by spronovo** [comment]
Hi, there are no plan for this. You should think about WSLg as a desktop environment (really Weston) with individual application integrated back into Windows. You can run app inside of WSLg, but running another desktop environment would really mean replacing WSLg altogether.  Well, in theory a desktop environment could have a Wayland backend and daisy chain to WSLg. For example Weston has a Wayland backend to you can run Weston on Weston... multiple time :-).

![image](https://user-images.githubusercontent.com/7545376/115980212-b6016700-a53f-11eb-9f2e-48c385f9a87f.png)

But the overall experience isn't going to give you what you are looking for.

If you want gnome as you desktop environment, you could run native Ubuntu, run Windows in a VM, and use something like FreeRDP to RAIL your Windows application back onto your Linux native desktop (essentially a sort of reverse WSLg).

**Reply by cerebrate** [reply]
I agree where the whole environment is concerned, but it would be useful if we could coax some bits of the Gnome desktop environment (or other desktop environments) into working under WSLg. (Disclaimer: This may be something that I'm thinking of mostly because of my own hacking on getting systemd/dbus up and running, dependency-wise.)

I'm thinking of three things, primarily. The first is gnome-keyring-daemon, which I have working more or less satisfactorily.

The second are policykit prompts, and getting them to show up via WSLg. I noted this one when WSLg created a shortcut for Synaptic Package Manager for me which invokes it as synaptic-pkexec - i.e., via pkexec - as it has to be run as root to actually manage packages, and this shortcut fails to work because pkexec can't find the graphical session, but it's certainly not the only app to need policykit prompting.

The third is notifications, as something else that various applications expect to work and which would be useful. I've tried notification-daemon and a couple of alternatives, but they each and every one segfault on startup, which at a guess is a side-effect of them not being able to find the graphical session either. (Actually, I find myself thinking that a good way to handle dbus notifications would be transmute them into Windows notifications so we only have one notification center to track, but hacking that project together is definitely one for my Copious Free Time.)

Any of this sound reasonable enough to maybe turn into a feature request? 😁

**Reply by spronovo** [reply]
Yeah that makes sense. These are not area I'm very familiar with yet, but the scenario make sense. For notification i agree we would like to eventually shuffle those over to Windows so they can be integrated with the rest of notifications. This is an area we need to dig more into and see if we can get pieces of these desktop environment to integrate better with WSLg. Do you know if there is a doc or overview somewhere how these pieces interconnect and work? (or I just need to read the source code lol).

**Reply by cerebrate** [reply]
Unfortunately, I'm still digging into a lot of that myself, starting from the casual-user perspective. The best I've found so far is https://www.freedesktop.org/software/polkit/docs/latest/ for PolicyKit and https://people.gnome.org/~mccann/docs/notification-spec/notification-spec-latest.html for notifications, but that's not a whole lot, alas.

**Reply by Guarmanda** [reply]
As cerebrate said,  these is a doc. But if there's anything lacking in it, you can always speak directly to some gnome devs, or find a projet where someone already took gnome notifications to take them somewhere else. Someone must have done a gnome extension about this, after all

**Reply by cerebrate** [reply]
It occurred to me that there's also gvfs, but it seems to work fine if you get a systemd user session up and running, so not a lot to do there.

**Comment by Guarmanda** [comment]
Later yesterday, I discovered that you where using Weston, and I managed, in fact, to use Weston on weston :D (the only Problem is I have two screen, but while wslg is loading the two, there is only 1 wayland screen so I can't force weston to start on the second one...)

You're right, anyway, dual boot or Windows vm on Linux is still better

**Comment by TobiaszCudnik** [comment]
Id love to run a full KDE/kwin desktop on top of windows, which would handle all the hardware - eg surface pro has some hardware issues on linux (power management, usbc, camera). Also as a compatibility layer for wayland (eg synergy / barrier doesnt work). Its a pity you're only planning a per-app mode for WSL2g. Cool job anyway!

**Reply by apos** [reply]
There are no problems, but solutions:

My workaround (but unfortunately it is not stable amoung hibernation / standby).

1. Start powershell or - better - the Linux distribution of your choice. 
2. Within the distirbution I the following to .bashrc:

The code uses

-  xfce4
- tint2 
- a custom script to get a temporary .X11-unix file within the session
- systemd (!) - please search in the forum how to install (if I remember right, I used: https://github.com/FiestaLake/ubuntu-wsl2-systemd-script) 

It starts gnome-session via **systemd** (!).

**.bashrc** 

```
# START XFCE4 environment with gnome-session
if ps x | grep -v grep | grep -v $$ | grep tint2 > /dev/null
then
        # do nothing
        cd ~
else

        cd ~/bin

        ./restore-x11-unix_wsl2.sh

        ps aux | grep -v grep | grep gnome-session || gnome-session --systemd &
        sleep 2

        sleep 5
        cd ~
fi
```

**restore-x11-unix_wsl2.sh**

```
#!/bin/bash
if [ -h /tmp/.X11-unix ]
then
        # echo "temporary link for X11 alreade set"
        exit 0
else

        ls -la /tmp/.X11-*
        sudo rm -r /tmp/.X11-unix
        ln -s /mnt/wslg/.X11-unix /tmp/.X11-unix

fi
```
3. Install tint2 within Ubuntu - it gives the best performance with Weston. I tried others (docky, planck ...) - they where not good to use. 

4. Exit and restart Ubuntu with the powershell. For future usage, you can just start Ubuntu with a powershell. [dot]Bahsrc will be executed. Systemd should now run automatically. 

5. Start **tint2 from windows start menu (!) - not from within the actual shell of the Ubuntu-System** . Then start ALL PROGRAMS FROM THIS PANEL - NOT THROUGH WINDOWS STARTMENU. You will be asked for a password, if you are starting a program that is bundled with gnome-keyring (e.g. evolution). This way all programms are using gnome-session if needed.

Cheers Axel


**Bugs:**

1. I encounter massive problems with copy and paste. This leads to a total crash running linux programs. Solution: you have to do "wsl shutdown" (unfortunately) an restart wsl.
2. Sometimes the programs simply disappear (tint2) because gnome-session seems to crash. But the linux session keeps going on and you can just start tint2 again.

**Reply by Guarmanda** [reply]
I tried your solution but... it did nothing^^ Like, it really asks for my password on startup but... nothing shows, not even errors^^
Thanks for yout tutorial anyway, it's very clear

**Reply by apos** [reply]
Hi @Guarmanda , 

I hope I did not miss anything. I am highly interested, the we find a reliable solution for this as long as we cannot use Ubuntu "as is". My reason: I like to use Evolution as E-Mail client. And I have a very good, old (and reliable) Ubuntu Installation. I used to use X410, but why use this overhead when we finally have a solution. 

You need to start Ubuntu with the .bashrc (the second script must be in the paht environment!) either with powershell or Windows Terminal. You can also merge the two scripts into one bashrc.. 

1. Verify, that gnome-session started: after Ubuntu started you should see something like:
/usr/libexec/gnome-session-binary --systemd --systemd
/usr/libexec/gnome-session-check-accelerated

2. if not: check, if systemd really works in Ubuntu:
`ps aux | grep systemd | grep "\-\-unit"`
root           1  0.0  0.0 172956 13100 ?        Ss   18:28   0:00 /lib/systemd/systemd --unit=multi-user.target


3. check, if the wslg socket ist connected:
`ls -la /tmp/.X11-unix`
lrwxrwxrwx 1 apos apos 19 Okt 29 18:35 /tmp/.X11-unix -> /mnt/wslg/.X11-unix

4. Then you should be able to start xterm from startmenu. If not, we should further check. 

Cheers Axel

**Reply by apos** [reply]
Post scriptum: tint2 can be a little bit buggy on multi monitor environments. To get it into correct position (e.g. button of the screen) can be tricky. Problably you just don't see it. Try planck or any other for testing purposes.

**Reply by Hephaestus-Dev** [reply]
Someone needs to post a video on Youtube very HOT TOPIC as this has been since wsl2 google Frankinbuntu or Winbuntu or Ubundows. Chris Titus did a live stream a while back I remember. ( https://www.youtube.com/watch?v=zDmSElHUiX0&t=4444s )
Please Delete this reply if necessary  : NOTE:  Above //  restore-x11-unix_wsl2.sh check spelling in syntax 
then
        # echo "temporary link for X11 alreade set"
        exit 0 
"already" ???
